junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt

```diff
@@ -158,8 +158,12 @@ inline fun <reified T : Throwable> assertThrows(noinline message: () -> String,
  * @param R the result type of the [executable]
  */
 @API(status = EXPERIMENTAL, since = "5.5")
-fun <R> assertDoesNotThrow(executable: () -> R): R =
-    Assertions.assertDoesNotThrow(ThrowingSupplier(executable))
+inline fun <reified R> assertDoesNotThrow(executable: () -> R): R = try {
+    val result = executable()
+    Assertions.assertDoesNotThrow(ThrowingSupplier { result })
+} catch (throwable: Throwable) {
+    Assertions.assertDoesNotThrow(ThrowingSupplier { throw throwable })
+}
```

Comment:
This might be neater?

```diff
diff --git a/junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt b/junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt
index cf420e7f3f..abe4e23769 100644
--- a/junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt
+++ b/junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt
@@ -158,11 +158,19 @@ inline fun <reified T : Throwable> assertThrows(noinline message: () -> String,
  * @param R the result type of the [executable]
  */
 @API(status = EXPERIMENTAL, since = "5.5")
-inline fun <reified R> assertDoesNotThrow(executable: () -> R): R = try {
-    val result = executable()
-    Assertions.assertDoesNotThrow(ThrowingSupplier { result })
-} catch (throwable: Throwable) {
-    Assertions.assertDoesNotThrow(ThrowingSupplier { throw throwable })
+inline fun <reified R> assertDoesNotThrow(executable: () -> R): R {
+    val (result, throwable) = try {
+        Pair(executable(), null)
+    } catch (throwable: Throwable) {
+        Pair(null as R, throwable)
+    }
+
+    return Assertions.assertDoesNotThrow(ThrowingSupplier {
+        if (throwable != null) {
+            throw throwable
+        }
+        result
+    })
 }
 
 /**
@@ -190,15 +198,20 @@ inline fun <reified R> assertDoesNotThrow(message: String, executable: () -> R):
  * @param R the result type of the [executable]
  */
 @API(status = EXPERIMENTAL, since = "5.5")
-inline fun <reified R> assertDoesNotThrow(noinline message: () -> String, executable: () -> R): R = try {
-    val result = executable()
-    Assertions.assertDoesNotThrow(
-        ThrowingSupplier { result },
-        Supplier(message)
-    )
-} catch (throwable: Throwable) {
-    Assertions.assertDoesNotThrow(
-        ThrowingSupplier { throw throwable },
+inline fun <reified R> assertDoesNotThrow(noinline message: () -> String, executable: () -> R): R {
+    val (result, throwable) = try {
+        Pair(executable(), null)
+    } catch (throwable: Throwable) {
+        Pair(null as R, throwable)
+    }
+
+    return Assertions.assertDoesNotThrow(
+        ThrowingSupplier {
+            if (throwable != null) {
+                throw throwable
+            }
+            result
+        },
         Supplier(message)
     )
 }
```

Reply:
I agree that the current implementation is not ideal because it calls `assertDoesNotThrow` twice, but I still find it more readable... might be a matter of personal preference though

Reply:
@williamboman Would something like this make sense?

```kotlin
inline fun <reified R> assertDoesNotThrow(executable: () -> R): R {
    Assertions.assertDoesNotThrow(getThrowingSupplier { executable })
}

inline fun <reified R> assertDoesNotThrow(noinline message: () -> String, executable: () -> R): R {
    Assertions.assertDoesNotThrow(
        getThrowingSupplier { executable },
        Supplier(message)
    )
}

inline fun <reified R> getThrowingSupplier(executable: () -> R): ThrowingSupplier<R> = try {
    val result = executable()
    ThrowingSupplier { result }
} catch (throwable: Throwable) {
    ThrowingSupplier { throw throwable }
}
```

Reply:
Yep I was imagining something similar ! Added in https://github.com/junit-team/junit5/pull/2684/commits/14407dd63b71b4b3b38b62da37ae548845bcda39

Reply:
Looks good! 👌